cabal-install/Distribution/Client/CmdSdist.hs

```diff
@@ -0,0 +1,289 @@
+    , sdistOutputPath    = mempty
+    }
+
+--
+
+sdistAction :: SdistFlags -> [String] -> GlobalFlags -> IO ()
+sdistAction SdistFlags{..} targetStrings globalFlags = do
+    let verbosity = fromFlagOrDefault normal sdistVerbosity
+        mDistDirectory = flagToMaybe sdistDistDir
+        mProjectFile = flagToMaybe sdistProjectFile
+        globalConfig = globalConfigFile globalFlags
+        listSources = fromFlagOrDefault False sdistListSources
+        archiveFormat = fromFlagOrDefault TargzFormat sdistArchiveFormat
+        mOutputPath = flagToMaybe sdistOutputPath
+  
+    projectRoot <- either throwIO return =<< findProjectRoot Nothing mProjectFile
+    let distLayout = defaultDistDirLayout projectRoot mDistDirectory
+    dir <- getCurrentDirectory
+    projectConfig <- runRebuild dir $ readProjectConfig verbosity globalConfig distLayout
+    baseCtx <- establishProjectBaseContext verbosity projectConfig
+    let localPkgs = localPackages baseCtx
+
+    targetSelectors <- either (reportTargetSelectorProblems verbosity) return
+        =<< readTargetSelectors localPkgs targetStrings
+    
+    let 
+        ext = case archiveFormat of
+            TargzFormat -> "tar.gz"
+            ZipFormat -> "zip"
+        
+        outputPath pkg = case mOutputPath of
+            Just path
+                | listSources -> path </> prettyShow (packageId pkg) <.> "list"
+                | otherwise   -> path </> prettyShow (packageId pkg) <.> ext
+            Nothing
+                | listSources -> "-"
+                | otherwise   -> distSdistFile distLayout (packageId pkg) archiveFormat
+
+    createDirectoryIfMissing True (distSdistDirectory distLayout)
+    
+    case reifyTargetSelectors localPkgs targetSelectors of
+        Left errs -> die' verbosity . unlines . fmap renderTargetProblem $ errs
+        Right pkgs -> mapM_ (\pkg -> packageToSdist verbosity listSources archiveFormat (outputPath pkg) pkg) pkgs
+
+data IsExec = Exec | NoExec
+            deriving (Show, Eq)
+
+packageToSdist :: Verbosity -> Bool -> ArchiveFormat -> FilePath -> UnresolvedSourcePackage -> IO ()
+packageToSdist verbosity listSources archiveFormat outputFile pkg = do
+    dir <- case packageSource pkg of
+        LocalUnpackedPackage path -> return path
+        _ -> die' verbosity "The impossible happened: a local package isn't local"
+    setCurrentDirectory dir
+
+    let norm flag = fmap ((flag, ) . normalise)
+    (norm NoExec -> nonexec, norm Exec -> exec) <- 
+        listPackageSources verbosity (flattenPackageDescription $ packageDescription pkg) knownSuffixHandlers
+
+    let write = if outputFile == "-" then BSL.putStrLn else BSL.writeFile outputFile
+        files =  nub . sortBy (\(_, a) (_, b) -> compare a b) $ nonexec ++ exec
+
+    if 
+        | listSources -> do
+            notice verbosity $ "File manifest for package " ++ prettyShow (packageId pkg) ++ ":\n"
+            write (BSL.pack . unlines . fmap snd $ files)
```

Comment:
`--list-only` is producing an extra newline at the end of its output, e.g.:

```
~/src/cabal$ cabal new-run cabal -- new-sdist Cabal --list-only
[...]
tests/custom-setup/IdrisSetup.hs
tests/hackage/check.sh
tests/hackage/download.sh
tests/hackage/unpack.sh
tests/misc/ghc-supported-languages.hs

~/src/cabal$ 
```

Reply:
Whoops, why is it `putStrLn`?